pkg/prom/instance/instance.go

```diff
@@ -306,47 +291,115 @@ func (i *Instance) Run(ctx context.Context) error {
+		if err != nil {
+			level.Error(i.logger).Log("msg", "failed to get scrape manager")
+			return err
+		}
+
 		// Scrape manager
 		rg.Add(
 			func() error {
-				err := scrapeManager.Run(discovery.SyncCh())
+				err := sm.Run(i.discovery.SyncCh())
 				level.Info(i.logger).Log("msg", "scrape manager stopped")
 				return err
 			},
 			func(err error) {
 				// The scrape manager is closed first to allow us to write staleness
 				// markers without receiving new samples from scraping in the meantime.
 				level.Info(i.logger).Log("msg", "stopping scrape manager...")
-				scrapeManager.Stop()
+				sm.Stop()
 
 				// On a graceful shutdown, write staleness markers. If something went
 				// wrong, then the instance will be relaunched.
 				if err == nil && i.cfg.WriteStaleOnShutdown {
 					level.Info(i.logger).Log("msg", "writing staleness markers...")
-					err := wstore.WriteStalenessMarkers(i.getRemoteWriteTimestamp)
+					err := i.wal.WriteStalenessMarkers(i.getRemoteWriteTimestamp)
 					if err != nil {
 						level.Error(i.logger).Log("msg", "error writing staleness markers", "err", err)
 					}
 				}
 
 				level.Info(i.logger).Log("msg", "closing storage...")
-				if err := storage.Close(); err != nil {
+				if err := i.storage.Close(); err != nil {
 					level.Error(i.logger).Log("msg", "error stopping storage", "err", err)
 				}
 			},
 		)
 	}
 
-	err = rg.Run()
+	err := rg.Run()
 	if err != nil {
 		level.Error(i.logger).Log("msg", "agent instance stopped with error", "err", err)
 	}
 	return err
 }
 
+// initialize sets up the various Prometheus components with their initial
+// settings. initialize will be called each time the Instance is run. Prometheus
+// components cannot be reused after they are stopped so we need to recreate them
+// each run.
+//
+// Components are stored as struct members to allow other methods to access
+// components from different goroutines (as opposed to keeping them completely
+// defined within the Run method where nothing else can access them).
+func (i *Instance) initialize(ctx context.Context, reg prometheus.Registerer) error {
+	i.componentMtx.Lock()
+	defer i.componentMtx.Unlock()
+
+	var err error
+
+	i.wal, err = i.newWal(reg)
+	if err != nil {
+		return fmt.Errorf("error creating WAL: %w", err)
+	}
+
+	i.discovery, err = i.newDiscoveryManager(ctx)
+	if err != nil {
+		return fmt.Errorf("error creating discovery manager: %w", err)
+	}
+
+	i.readyScrapeManager = &scrape.ReadyScrapeManager{}
+
+	// Setup the remote storage
+	remoteLogger := log.With(i.logger, "component", "remote")
+	i.remoteStore = remote.NewStorage(remoteLogger, reg, i.wal.StartTime, i.wal.Directory(), i.cfg.RemoteFlushDeadline, i.readyScrapeManager)
+	err = i.remoteStore.ApplyConfig(&config.Config{
+		GlobalConfig:       i.globalCfg,
+		RemoteWriteConfigs: i.cfg.RemoteWrite,
+	})
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed applying config to remote storage", "err", err)
```

Comment:
Do we need these log lines? From what I read the error will bubble up and be logged at the same log level.